apps/labs/posts/libsf-error-state.md

```diff
@@ -0,0 +1,744 @@
+    SF_ERROR_IGNORE, /* SF_ERROR_UNDERFLOW */
+    SF_ERROR_IGNORE, /* SF_ERROR_OVERFLOW */
+    SF_ERROR_IGNORE, /* SF_ERROR_SLOW */
+    SF_ERROR_IGNORE, /* SF_ERROR_LOSS */
+    SF_ERROR_IGNORE, /* SF_ERROR_NO_RESULT */
+    SF_ERROR_IGNORE, /* SF_ERROR_DOMAIN */
+    SF_ERROR_IGNORE, /* SF_ERROR_ARG */
+    SF_ERROR_IGNORE, /* SF_ERROR_OTHER */
+    SF_ERROR_IGNORE  /* SF_ERROR__LAST */
+};
+```
+
+To make the correspondence between entries in the array and special function error handling types clear, there is
+another enumeration type, `sf_error_t`, which associates the name for each error type with the corresponding index into
+`sf_error_actions`.
+
+```c
+typedef enum {
+    SF_ERROR_OK = 0,    /* no error */
+    SF_ERROR_SINGULAR,  /* singularity encountered */
+    SF_ERROR_UNDERFLOW, /* floating point underflow */
+    SF_ERROR_OVERFLOW,  /* floating point overflow */
+    SF_ERROR_SLOW,      /* too many iterations required */
+    SF_ERROR_LOSS,      /* loss of precision */
+    SF_ERROR_NO_RESULT, /* no result obtained */
+    SF_ERROR_DOMAIN,    /* out of domain */
+    SF_ERROR_ARG,       /* invalid input parameter */
+    SF_ERROR_OTHER,     /* unclassified error */
+    SF_ERROR__LAST
+} sf_error_t;
+```
+
+`sf_action_t` is declared static to make use of its name local only to the file where it's defined. To work with this
+array externally, there are two functions, `scipy_sf_error_set_action` for updating and `scipy_sf_error_get_action` for
+querying the error handling policy associated to a given error type.
+
+```c
+void scipy_sf_error_set_action(sf_error_t code, sf_action_t action)
+{
+	sf_error_actions[(int)code] = action;
+}
+
+sf_action_t scipy_sf_error_get_action(sf_error_t code)
+{
+    return sf_error_actions[(int)code];
+}
+```
+
+If a user runs the following in an IPython session
+
+```python
+In  [1]: import scipy.special as special
+
+In  [2]: special.seterr(loss="warn")
+```
+
+then somewhere down in the stack, the following call will occur to update the error state.
+
+```c
+scipy_sf_error_set_action(SF_ERROR_LOSS, SF_ERROR_WARN)
+```
+
+If `special.gamma` is evaluated at an array containing a negative integer, then when the scalar kernel is evaluating the
+offending entry, it will signal `SF_ERROR_SINGULAR` and `sf_error_actions` will be consulted to figure out what action
+to take.
+
+## The many battles to actually ship it
+
+The shared library `libsf_error_state`'s contents are fairly simple, but how does one actually ship a shared library
+within a Python package like SciPy? The process for creating and using shared libraries depends on the operating system
+and compiler toolchain. SciPy supports a wide range of platforms in aim of its goal to democratize access to scientific
+computing tools and the greatest challenge turned out to be getting things to work on each of them. Several times, just
+as we thought everything was finally working, another quirk would pop up that needed to be addressed.
```

Comment:
It may be worth pointing out that it's extremely rare for Python packages to contain an intenral shared library - we didn't know of other examples of this being done when we started working on this topic.